src/usexp/dune_lexer.mll

```diff
@@ -4,6 +4,91 @@ open Lexer_shared
 type block_string_line_kind =
   | With_escape_sequences
   | Raw
+
+module Template = struct
+  include Template
+
+  let dummy_loc =
+    { Loc.
+      start = Lexing.dummy_pos
+    ; stop = Lexing.dummy_pos
+    }
+
+  let add_text parts s =
+    match parts with
+    | Template.Text s' :: parts -> Template.Text (s' ^ s) :: parts
+    | _ -> Template.Text s :: parts
+
+  let token parts ~quoted ~start (lexbuf : Lexing.lexbuf) =
+    lexbuf.lex_start_p <- start;
+    match parts with
+    | [] | [Text ""] ->
+      error lexbuf "Internal error in the S-expression parser, \
+                    please report upstream."
+    | [Text s] ->
+      Token.Atom (Atom.of_string s)
+    | _ ->
+      Token.Template
+        { quoted
+        ; loc = dummy_loc
+        ; parts = List.rev parts
+        }
+
+  module Buffer : sig
+    val new_token : unit -> unit
+    val get : unit -> Token.t
+    val add_var : part -> unit
+    val add_text : string -> unit
+    val add_text_c : char -> unit
+  end = struct
+    type state =
+      | String
+      | Template of Template.part list
+
+    let text_buf = Buffer.create 256
```

Comment:
I understand that lexers are mutable by nature, but what do you think about passing the buffer explicitly? E.g. define an abstract `Buffer.t` type, a constructor, and add a global `buffer : Buffer.t`.

Reply:
Sure, I can do that. I think @diml used a global variable like this for the escaped_buf so I thought I might as well do the same. I don't mind passing stuff around, but would prefer if it's done consistently everywhere.

Reply:
Yh, I haven't bother too much with global variables given that dune is single threaded and that we don't expose these libraries. If we ever made them public we should be careful about that. Here I don't mind whether we keep a global variable or not.